File: LeetCode/1-100/46. 全排列.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permute = function (nums) {
  const numsLen = nums.length
  const result = []
  const fn = (index) => {
    if (index === numsLen) {
      return result.push([...nums])
    }
    for (let i = index; i < numsLen; i++) {
      ;[nums[i], nums[index]] = [nums[index], nums[i]]
      fn(index + 1)
      ;[nums[index], nums[i]] = [nums[i], nums[index]]
    }
  }
  fn(0)
  return result
}

console.log(permute([1, 2, 3]))

/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permute2 = function (nums) {
  const res = [],
    path = []
  const backtrack = (n, k, used) => {
    if (path.length === k) {
      return res.push([...path])
    }
    for (let i = 0; i < k; ++i) {
      if (used[i]) continue
      used[i] = true
      path.push(n[i])
      backtrack(n, k, used)
      used[i] = false
      path.pop()
    }
  }
  backtrack(nums, nums.length, [])
  return res
}

File: LeetCode/1-100/46. 全排列.java
class Solution {
   boolean[] used;
    List<List<Integer>> an2 = new ArrayList<>();

    public List<List<Integer>> permute(int[] nums) {
        int n = nums.length;
        used = new boolean[n];
        dfs2(nums,new ArrayList<>(),n);
        return an2;
    }
    public void dfs2(int[] nums,List<Integer> cur, int k) {
        if (cur.size() == k) {
            an2.add(new ArrayList<>(cur));
            return;
        }
        for (int i = 0; i < k; i++) {
            if (used[i]) continue;
            used[i] = true;
            cur.add(nums[i]);
            dfs2(nums,cur,k);
            cur.remove(cur.size() - 1);
            used[i] = false;
        }
    }
}